mypy/checkexpr.py

```diff
@@ -2782,8 +2805,12 @@ def has_shape(typ: Type) -> bool:
             formal_to_actual = map_actuals_to_formals(
                 arg_kinds, arg_names, typ.arg_kinds, typ.arg_names, lambda i: arg_types[i]
             )
-
             with self.msg.filter_errors():
+                if typ.is_generic() and typ.param_spec() is not None:
+                    typ, formal_to_actual = self.adjust_generic_callable_params_mapping(
+                        typ, args, arg_kinds, arg_names, formal_to_actual, context
+                    )
```

Comment:
I'm confused about why this depends on `typ.param_spec()` when it doesn't in `check_callable_call`. Removing that makes sense to me but also fails a ton of test cases. Do you know why?

Reply:
Plain typevars are irrelevant here: this function looks for targets that have proper arguments **count** (signature shape, ignoring parameters types). Typevar resolution cannot affect this shape since it's a 1:1 mapping. ParamSpec, on the other hand, might be expanded into some fixed-arg form if we can solve this ParamSpec now. 

Removing this constraint results in a bunch of overload resolution failures, like

```
from typing import Any, Callable, overload, TypeVar

F = TypeVar('F', bound=Callable[..., Any])

@overload
def dec(x: F) -> F: ...
@overload
def dec(x: str) -> Callable[[F], F]: ...
def dec(x) -> Any:
    pass

@dec
def f(name: str) -> int:
    return 0

@dec('abc')
def g(name: str) -> int:
    return 0

reveal_type(f)  # N: Revealed type is "def (name: builtins.str) -> builtins.int"
reveal_type(g)  # N: Revealed type is "def (name: builtins.str) -> builtins.int"
```

If we try to expand, `dec("abc")` begins to match the first overload, and then "str not callable". That's because `infer_function_type_arguments` is too lax and does not include some constraints - I don't know whether it's intended, but sounds difficult to fix. Such impossible resolutions are filtered out later in `check_callable_call`, so it doesn't cause any false negatives with regular calls - but `plausible_overload_call_targets` does not need such precision, we're just filtering out overloads that certainly can't match.

This piece supports scenarios similar to `testRunParamSpecOverload` testcase - without such check, ParamSpec overloads will be unexpectedly filtered out. As an alternative, we can avoid this expansion entirely and just report a match whenever paramspec is encountered in `overload` - that may be even more robust, actually.

Reply:
Hm, and that alternative also passes all testcases. I think it's a safer solution that won't harm: the most severe impact of adding an extra overload would be a performance overhead to check one. Overload+paramspec isn't a very popular combination, so the performance impact shouldn't be noticeable. I pushed a version with this update, let's see what CI says.